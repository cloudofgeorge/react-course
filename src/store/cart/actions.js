import { CART_ADD_ITEM, CART_REMOVE_ITEM } from "./types";

export const addCartItemAction = (data) => ({
  type: CART_ADD_ITEM,
  payload: data,
});

export const removeCartItemAction = (data) => ({
  type: CART_REMOVE_ITEM,
  payload: data,
});
