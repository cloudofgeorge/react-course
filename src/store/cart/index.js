export { cartReducer } from "./reducer";
export * from "./actions";
export * from "./selectors";
