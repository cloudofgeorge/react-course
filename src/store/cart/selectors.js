import { createSelector } from 'reselect';

const stateCart = state => state.cart;

export const getCartData = createSelector(stateCart, state => state.data);
