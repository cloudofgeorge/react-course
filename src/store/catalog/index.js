export { catalogReducer } from "./reducer";
export * from "./actions";
export * from "./selectors";
