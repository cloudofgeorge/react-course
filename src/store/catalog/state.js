export const initialState = {
  isFetching: false,
  data: [],
  error: null,
};
