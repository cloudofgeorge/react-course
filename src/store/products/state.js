export const initialState = {
  pending: false,
  fetched: false,
  data: {
    products: 0,
    cart: [],
  },
  error: null,
};
