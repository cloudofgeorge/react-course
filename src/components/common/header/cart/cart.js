import React, { useCallback, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import styles from './cart.module.scss';
import { removeCartItemAction, getCartData } from '../../../../store/cart';

export const Cart = () => {
  const [isShow, setIsShow] = useState(false);

  const cartData = useSelector(getCartData);

  const dispatch = useDispatch();
  const removeCartItem = useCallback(item => dispatch(removeCartItemAction(item)), [dispatch]);

  const toggleWindow = () => {
    if (cartData && cartData.length) {
      setIsShow(!isShow);
    }
  };

  return (
    <div className={styles.cart}>
      <button onClick={toggleWindow} disabled={cartData.length === 0} data-testid="cart-button">
        cart {cartData.length > 0 && cartData.length}
      </button>

      {isShow && cartData.length > 0 && (
        <div className={styles.cartWindow} data-testid="cart-window">
          {cartData.map(item => (
            <div key={item.id} className={styles.cartItem} data-testid="cart-item">
              <span className={styles.cartItemName}>{item.title}</span>
              <button onClick={() => removeCartItem(item)} data-testid="cart-remove-item-button">
                x
              </button>
            </div>
          ))}
        </div>
      )}
    </div>
  );
};
