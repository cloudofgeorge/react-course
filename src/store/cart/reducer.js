import { initialState } from "./state";
import { CART_ADD_ITEM, CART_REMOVE_ITEM } from "./types";

export const cartReducer = (state = initialState, action) => {
  switch (action.type) {
    case CART_ADD_ITEM: {
      return { data: [...state.data, action.payload] };
    }

    case CART_REMOVE_ITEM: {
      return { data: state.data.filter(({ id }) => id !== action.payload.id) };
    }

    default:
      return state;
  }
};
