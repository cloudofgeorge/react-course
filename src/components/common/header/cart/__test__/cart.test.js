import React from 'react';
import { fireEvent } from '@testing-library/react';
import { Cart } from '../cart';
import { renderWithRedux } from '../../../../../test/utils';
import { state } from './state';
import { getCartData } from '../../../../../store/cart';

jest.mock('../../../../../store/cart', () => ({
  getCartData: jest.fn(),
}));

beforeEach(() => {
  getCartData.mockClear();
});

describe('<Cart>', () => {
  it('should match snapshot', () => {
    getCartData.mockImplementation(() => state);

    const result = renderWithRedux(<Cart />);
    expect(result.container).toMatchSnapshot();
  });

  it('should open popup', () => {
    getCartData.mockImplementation(() => state);

    const result = renderWithRedux(<Cart />);

    fireEvent.click(result.getByTestId('cart-button'));

    expect(result.getByTestId('cart-window')).toBeDefined();
    expect(result.getAllByTestId('cart-item').length).toEqual(15);
  });
});
