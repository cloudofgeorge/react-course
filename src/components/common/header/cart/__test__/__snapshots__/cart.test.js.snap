// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Cart> should match snapshot 1`] = `
<div>
  <div
    class="cart"
  >
    <button
      data-testid="cart-button"
    >
      cart 
      15
    </button>
  </div>
</div>
`;
